dynamic_instrumentation/src/instrument_memory.rs

```diff
@@ -236,16 +236,60 @@ impl<'a, 'tcx: 'a> CollectFunctionInstrumentationPoints<'a, 'tcx> {
         find_instrumentation_def(self.tcx, self.runtime_crate_did, name)
     }
 
-    fn should_instrument(&self, local: Local, context: PlaceContext) -> bool {
-        let is_unsafe_ptr = self.body.local_decls[local].ty.is_unsafe_ptr();
-        is_unsafe_ptr && context.is_use()
+    fn is_shared_or_unsafe_ptr(&self, ty: &TyS<'tcx>) -> bool {
+        ty.is_unsafe_ptr() || (ty.is_region_ptr() && !ty.is_mutable_ptr())
     }
 
     fn func_hash(&self) -> mir_loc::DefPathHash {
         self.tcx.def_path_hash(self.body.source.def_id()).convert()
     }
 }
 
+fn has_outer_deref(p: &Place) -> bool {
+    matches!(
+        p.iter_projections().last(),
+        Some((_, ProjectionElem::Deref))
+    )
+}
+
+fn pop_last_projection<'tcx>(p: &Place<'tcx>, tcx: TyCtxt<'tcx>) -> Option<Place<'tcx>> {
+    let Place { local, projection } = *p;
+    projection.split_last().map(|(_last, rest)| Place {
+        local,
+        projection: tcx.intern_place_elems(rest),
+    })
+}
+
+/// Get the inner-most dereferenced [`Place`].
+fn strip_all_deref<'tcx>(p: &Place<'tcx>, tcx: TyCtxt<'tcx>) -> Place<'tcx> {
+    let mut base_dest = *p;
+    let mut place_ref = base_dest.as_ref();
+    while let Some((cur_ref, proj)) = place_ref.last_projection() {
+        if let ProjectionElem::Deref = proj {
+            base_dest = Place {
+                local: cur_ref.local,
+                projection: tcx.intern_place_elems(cur_ref.projection),
+            };
+        }
+        place_ref = cur_ref;
+    }
+
+    base_dest
+}
+
+/// Used to strip initital deref from projection sequences
```

Comment:
```suggestion
Used to strip initital [`Deref`](ProjectionElem::Deref) from projection sequences.
```

Reply:
I think the comment string /// is gone in your suggested change

Reply:
whoops